app/src/main/java/org/simple/clinic/newentry/PatientEntryUpdate.kt

```diff
@@ -1,58 +1,71 @@
-
-      is DateOfBirthChanged -> Next.next(model.withDateOfBirth(event.dateOfBirth), setOf(HideDateOfBirthErrors))
-
-      is FullNameChanged -> Next.next(model.withFullName(event.fullName), setOf(ShowEmptyFullNameError(false)))
-
-      is PhoneNumberChanged -> Next.next(model.withPhoneNumber(event.phoneNumber), setOf(HidePhoneLengthErrors))
-
-      is ColonyOrVillageChanged -> Next.next(model.withColonyOrVillage(event.colonyOrVillage), setOf(HideEmptyColonyOrVillageError))
+      is OngoingEntryFetched -> onOngoingEntryFetched(model, event.patientEntry)
+      is GenderChanged -> onGenderChanged(model, event.gender)
+      is AgeChanged -> onTextChanged(HideEmptyDateOfBirthAndAgeError) { model.withAge(event.age) }
+      is DateOfBirthChanged -> onTextChanged(HideDateOfBirthErrors) { model.withDateOfBirth(event.dateOfBirth) }
+      is FullNameChanged -> onTextChanged(ShowEmptyFullNameError(false)) { model.withFullName(event.fullName) }
+      is PhoneNumberChanged -> onTextChanged(HidePhoneLengthErrors) { model.withPhoneNumber(event.phoneNumber) }
+      is ColonyOrVillageChanged -> onTextChanged(HideEmptyColonyOrVillageError) { model.withColonyOrVillage(event.colonyOrVillage) }
+      is DistrictChanged -> onTextChanged(HideEmptyDistrictError) { model.withDistrict(event.district) }
+      is StateChanged -> onTextChanged(HideEmptyStateError) { model.withState(event.state) }
+      is DateOfBirthFocusChanged -> onDateOfBirthFocusChanged(model, event.hasFocus)
+      is SaveClicked -> onSaveClicked(model.patientEntry!!)
+      PatientEntrySaved -> justEffect(OpenMedicalHistoryEntryScreen)
+    }
+  }
 
-      is DistrictChanged -> Next.next(model.withDistrict(event.district), setOf(HideEmptyDistrictError))
+  private fun onOngoingEntryFetched(
+      model: PatientEntryModel,
+      patientEntry: OngoingNewPatientEntry
+  ): PatientEntryNext =
+      next(model.patientEntryFetched(patientEntry), PrefillFields(patientEntry))
 
-      is StateChanged -> Next.next(model.withState(event.state), setOf(HideEmptyStateError))
+  private fun onGenderChanged(model: PatientEntryModel, gender: Optional<Gender>): PatientEntryNext {
+    val updatedModel = model.withGender(gender)
+    return if (gender.isNotEmpty() && model.isSelectingGenderForTheFirstTime) {
+      next(updatedModel.selectedGender(), setOf(HideMissingGenderError, ScrollFormToBottom))
+    } else {
+      next(updatedModel)
+    }
+  }
 
-      is DateOfBirthFocusChanged -> {
-        val hasDateOfBirth = model.patientEntry?.personalDetails?.dateOfBirth?.isNotBlank() == true
-        // TODO(rj): 2019-10-04 Extract the justEffect function and use it instead.
-        Next.dispatch<PatientEntryModel, PatientEntryEffect>(setOf(ShowDatePatternInDateOfBirthLabel(event.hasFocus || hasDateOfBirth)))
-      }
+  private fun onTextChanged(effect: PatientEntryEffect, reducer: () -> PatientEntryModel): PatientEntryNext =
+      next(reducer(), effect)
```

Comment:
The reducer doesn't actually do anything here. Why can't we just pass in the updated model directly?

```kotlin
onTextChanged(HideEmptyDateOfBirthAndAgeError, model.withAge(event.age))
```